### DISCUSSION TITLE
user script only run for 1st url of each tab

### DISCUSSION THREAD

**Discussion by n-buna404** [bromite/bromite#2137] [Q&A] [answered]
I had written a user script to remove open-inapp button at bottom middle of screen
for the below website
example website: https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7151093
this is the user script sample. 
[屏蔽開啓 APP.txt](https://github.com/bromite/bromite/files/8889516/APP.txt)
I test on chrome pc using tampermonkey and changing user agent to mobile. which the user script works fine in removing the button except being slightly sluggish

but when i tested in bromite, the button will only be blocked first time only after bromite opened but not subsequent pages.
I tested with both JIT enabled and disabled but there is no difference still only 1st page after opening bromite will execute the script successfully 
Is there any reason why this behavior occurs? I did not report this as bug as it may be due to my problematic script which i learnt to write in a few hours only.

**Comment by n-buna404** [comment]
After testing with tampermonkey on firefox browser(nightly), I found the discrepancy in behaviour of user script manager of bromite and tampermonkey on firefox. Bromite user script manager will only run the script **only for the 1st url of each tab** regardless of page reload or redirecting whereas tampermonkey will rerun the script every time page reload. Is this an intended behavior?

Methodology
testing script via scrolling in https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=3282690&last=
scrolling will remove the open app button if user script executed.
click other recommended posts on the bottom and it will redirect to other post then scroll again.

Script
```
> // @match        https://m.gamer.com.tw/forum/*
// @grant        none
// ==/UserScript==
window.addEventListener('scroll', function() {
    const elements = [...document.querySelectorAll('.open-app')];
elements.forEach(e => e.classList.remove('open-app'));
}, false);
```

**Reply by n-buna404** [reply]
@uazo would appreciate any clarification on the userscript manager behavior since you authored the pull request for the feature.

**Reply by uazo** [reply]
>  Is this an intended behavior?

No, you could try with a simple `alert`?

**Comment by n-buna404** [comment]
Yup i tested with simple alert. When i used  navigation key to move to next
page of example website
https://m.gamer.com.tw/forum/C.php?bsn=36726&snA=184&bpage=1&ltype=&page=1
the scroll to alert is not functioning.
This differ from tamper monkey behavior. As i pressed next page then the
scroll to alert still function

// ==/UserScript==
*window.addEventListener*('scroll', function() {
    alert('test');
}, false);

**Reply by uazo** [reply]
what I was asking is the test with a script where the `alert` was not inside `addEventListener`, simply to show you that the script is called on every page load.

> This differ from tamper monkey behavior

the behavior in bromite may be different, there was never a will to make bromite userscript work like tampermonkey

**Reply by n-buna404** [reply]
> what I was asking is the test with a script where the `alert` was not inside `addEventListener`, simply to show you that the script is called on every page load.
 
Okay i used only 
> alert('test');
here is the video for demonstration since it might be my poor communication skills and understanding
https://www107.zippyshare.com/v/EaO8qaW3/file.html

Is bromite user script manager considered called on every page load from ur perspective?
If yes I would consider that as normal behaviour of bromite userscript manager and marked this as solved.

**Reply by uazo** [reply]
> https://www107.zippyshare.com/v/EaO8qaW3/file.html

that url doesn't work

> called on every page load

bromite injects userscripts into each page according to what was defined with `@run-at`, see https://github.com/bromite/bromite/wiki/UserScripts

**Reply by n-buna404** [reply]
> > https://www107.zippyshare.com/v/EaO8qaW3/file.html
> 
> that url doesn't work
> 
> > called on every page load
> 
> bromite injects userscripts into each page according to what was defined with `@run-at`, see https://github.com/bromite/bromite/wiki/UserScripts

link works at my end but here another site https://anonfiles.com/DaweYeq3y1/20220620_200853_mp4

**Reply by uazo** [reply]
please try with this:
```
// ==UserScript==
// @name alert1_doument-start
// @version 0.1.00
// @description alert1
// @author mac
// @run-at document-start
// @match https://m.gamer.com.tw/forum/*
// ==/UserScript==
'use strict';
(() => alert('alert 1\nalert_doument-start'))();
```

**Reply by n-buna404** [reply]
> // ==UserScript==
> // @name alert1_doument-start
> // @version 0.1.00
> // @description alert1
> // @author mac
> // @run-at document-start
> // @match https://m.gamer.com.tw/forum/*
> // ==/UserScript==
> 'use strict';
> (() => alert('alert 1\nalert_doument-start'))();

i tested and it shows similar behavior as previous video
here is the new script demo: https://anonfiles.com/T192Y3q5y0/20220620_205943_mp4

edit: this can replicated for other forum site such as https://forum.lowyat.net/*
My guess is bromite userscript manager somehow thinks other forum post with slightly different url as same site and inject the script once only if the url change is not directly from urlbar but redirect via links

**Reply by uazo** [reply]
please enable `chrome://flags/#enable-userscripts-log` flag and send the log taken with `adb logcat`

**Reply by n-buna404** [reply]
[logcat.txt](https://github.com/bromite/bromite/files/8940891/logcat.txt)
here is the log. If u need any thing else please let me know.
edit: I did clicked for other recommended post to redirect but i cannot find the next url in log

**Reply by uazo** [reply]
here some relevant lines:
```
UserScripts: DidCreateDocumentElement -> DOCUMENT_START
UserScripts: GetAllInjections
UserScripts: No Match name=alert1_doument-start id=_6 url=https://api.gamer.com.tw/worker/signin.html
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
UserScripts: GetAllInjections
UserScripts: No Match name=alert1_doument-start id=_6 url=https://api.gamer.com.tw/worker/signin.html
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UserScripts: RunIdle -> DOCUMENT_IDLE
UserScripts: GetAllInjections
UserScripts: No Match name=alert1_doument-start id=_6 url=https://api.gamer.com.tw/worker/signin.html
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I'd like to understand why that url is indicated even though it doesn't appear in your video.
one of the possible explanations is that, being that url relative to a worker, probably the navigation at DOCUMENT_START is intercepted and appears as such. are you logged in to the site?

other interesting lines:
```
UserScripts: GetAllInjections
UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=6083674&last=
UserScripts: run location for name=alert1_doument-start id=_6 current 2 need 1
                                                              ^^^^^^^^^^^^^^^^
```
basically, I reads that if, instead of `@run-at document-start` (need **1**), you should use `@run-at document-end` (current **2**) so It should work

**Reply by n-buna404** [reply]
> 
Yes I am login into the website. My apologize for not releasing that it may causes error.
Now i tried with out login and cleared browsing data. also changed the @run-at document-end. However I still able to replicate this. 
Here is the log for it.
[logcat2.txt](https://github.com/bromite/bromite/files/8946957/logcat2.txt)
Is this behaviour not reproducible and your devices?
Another thing to my concern is 
>https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=6083674&last=

snA=6083674 should change to other number as I tap on link at bottom to redirect but it do not show inside the log. As it indicates the serial number of post (bsn represent the board serial number). For some reason the page change did happen but it was not logged?

**Reply by uazo** [reply]
see here:
```
UserScripts: DidCreateDocumentElement -> DOCUMENT_START
UserScripts: GetAllInjections
UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7168386&bpage=1&ltype=
UserScripts: run location for name=alert1_doument-start id=_6 current 1 need 2
UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
UserScripts: GetAllInjections
UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7168386&bpage=1&ltype=
UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7168386&bpage=1&ltype=
UserScripts: Injecting https://userscripts/file/POmA6Smk1/GKwc2WlwEh8Uj46rJxyHMGyuHxyaCOiJ0=.js
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UserScripts: RunIdle -> DOCUMENT_IDLE
UserScripts: GetAllInjections
UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7168386&bpage=1&ltype=
UserScripts: run location for name=alert1_doument-start id=_6 current 3 need 2
```
if once injected it doesn't work, it's because of the script

> My apologize for not releasing that it may causes error.

don't worry

> Is this behaviour not reproducible and your devices?

it works for me, but I tried without logging in (and, i'm sorry, no intention of creating a login)

**Reply by n-buna404** [reply]
> if once injected it doesn't work, it's because of the script

Correct me if I am wrong. What i understand here is script was not running because somehow it thought that the redirected page url is the same as original page so it will not run again.

**Reply by n-buna404** [reply]
>it works for me, but I tried without logging in (and, i'm sorry, no intention of creating a login)

I tested with another device specially samsung J1(2016) and cannot replicate the behaviour ( Userscript run on every redirect post). Then tried to install fresh bromite on Mi A1 and test again.However, it still do not run script when redirected. Surprise to see the issue is device specific.

**Reply by uazo** [reply]
if you can understand why, please let us know, I am curious.

**Reply by uazo** [reply]
> Correct me if I am wrong. What i understand here is script was not running because somehow it thought that the redirected page url is the same as original page so it will not run again.

no, not really for that.
I'm not sure, but I think that since a worker can be installed from the site (which basically intercepts the navigation, for example useful for managing offline), when verifying the url in DOCUMENT_START, the origin appears as shown in log rather than the actual one.
in chromium there are different types of origins, even opaque ones (like data urls), and still my understanding is not complete. if it happened on a site without having to create login I could understand better

**Reply by n-buna404** [reply]
Yes! I finally found root cause to the issue:
When add script to the userscript manager in bromite, I used Zarchiver to pinpoint the file location.
However, if I used Files build-in to the android system. The issue will not occur.
Thanks a lot @uazo for ur time and effort in helping me to pinpoint the problem.

**Reply by uazo** [reply]
I don't understand what it has to do with it.
by chance the content you see in the source is different based on what you choose to select the userscript?
if so, will you show me please?

**Reply by n-buna404** [reply]
> I don't understand what it has to do with it. by chance the content you see in the source is different based on what you choose to select the userscript? if so, will you show me please?

Yes u are correct I sorry that I jumped into conclusion too quickly.Basically because Bromite on J1(2016) is on 95.04638.79(last supported version for the device). I tried that version on Mi A1 and thought that different file add will cause problem (but it did not). 
After trial and error from each version from 95.04638.79. I found that 100.0.4896.135 was the last working with user script. 
I tested with one new tab, paste url and load once then redirect page 4 times and all 5 works. However starting 101.0.4951.39 out of 5 only 1st two instances which pasting and also 1st redirect works. The next 3 failed.
I hereby attached the log for v100 and v101
[logcatv100.txt](https://github.com/bromite/bromite/files/8948345/logcatv100.txt)
[logcatv101.txt](https://github.com/bromite/bromite/files/8948348/logcatv101.txt)

**Reply by n-buna404** [reply]
[logcatv100_2.txt](https://github.com/bromite/bromite/files/8948431/logcatv100_2.txt)
[logcatv101_3.txt](https://github.com/bromite/bromite/files/8948478/logcatv101_3.txt)

please ignore the earlier file as I forgot to turn userscript logging flag

edit: I am also curious if u r able to replicate this issue with total of 5 clicks as I jumped to conclusion when 1st redirect works but the subsequent failed. 
I successfully replicate this issue( open new tab, paste url>load and click links to redirect 4 times
only 1st 2 works) on 102.0.5005.96 arm non64 version Redmi note5Pro also.
all this performed without login.

**Reply by uazo** [reply]
based on the evidence of the logs you have provided I do not detect any problems, except a bug in the logging that I will fix:

```
06-21 19:30:57.345 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:30:57.370 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:30:57.370 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
06-21 19:30:57.370 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
06-21 19:30:57.370 30293 30308 I chromium: [INFO:user_script_set.cc(242)] UserScripts: Injecting https://userscripts/file/POmA6Smk1/GKwc2WlwEh8Uj46rJxyHMGyuHxyaCOiJ0=.js

06-21 19:31:02.414 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:31:02.428 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:31:02.428 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=
06-21 19:31:02.428 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=

06-21 19:31:07.575 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:31:07.584 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:31:07.584 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=
06-21 19:31:07.584 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=

06-21 19:31:11.688 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:31:11.693 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:31:11.693 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
06-21 19:31:11.693 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=

06-21 19:31:11.688 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:31:11.693 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:31:11.693 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
06-21 19:31:11.693 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=

06-21 19:31:15.870 30293 30308 I chromium: [INFO:script_injection_manager.cc(171)] UserScripts: DidDispatchDOMContentLoadedEvent -> DOCUMENT_END
06-21 19:31:15.895 30293 30308 I chromium: [INFO:user_script_set_manager.cc(44)] UserScripts: GetAllInjections
06-21 19:31:15.895 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=
06-21 19:31:15.895 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=
```
the error I find is that `UserScripts: Injecting` is written to the log only the first time, but the problem is in the log, I have no reason to believe that it is not injected correctly (because, if the execution enters [here](https://github.com/bromite/bromite/blob/6c2d2ee368490330c65276995714fde70f9f3bb2/build/patches/Experimental-user-scripts-support.patch#L9946) it is not logged).
if anyone else finds the same problem and can give me more information on how to reproduce it, I will be happy to fix it.

regarding the worker problem and the log bug, @csagan5 I will ask you to modify my patch with https://gist.github.com/uazo/c114f68d5c0c781a85045f958a8373da maybe after 103

**Reply by n-buna404** [reply]
v100.0.4896.135
`Line 5705: 06-21 19:30:56.132 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5733: 06-21 19:30:57.370 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5734: 06-21 19:30:57.370 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5750: 06-21 19:30:58.413 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5807: 06-21 19:31:01.789 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=
	Line 5817: 06-21 19:31:02.428 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=
	Line 5818: 06-21 19:31:02.428 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=
	Line 5825: 06-21 19:31:03.281 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7167439&last=
	Line 5871: 06-21 19:31:07.199 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=
	Line 5875: 06-21 19:31:07.584 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=
	Line 5876: 06-21 19:31:07.584 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=
	Line 5883: 06-21 19:31:08.439 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7166965&last=
	Line 5919: 06-21 19:31:11.209 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
	Line 5935: 06-21 19:31:11.693 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
	Line 5936: 06-21 19:31:11.693 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
	Line 5941: 06-21 19:31:12.511 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165926&last=
	Line 5984: 06-21 19:31:15.179 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=
	Line 5988: 06-21 19:31:15.895 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=
	Line 5989: 06-21 19:31:15.895 30293 30308 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=
	Line 5994: 06-21 19:31:16.778 30293 30308 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165826&last=`
v101.04951.39
`Line 5907: 06-21 19:40:09.180   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5911: 06-21 19:40:10.250   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5912: 06-21 19:40:10.250   497   512 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 5940: 06-21 19:40:11.368   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&snA=7167620
	Line 6112: 06-21 19:40:15.655   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165962&last=
	Line 6116: 06-21 19:40:15.981   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165962&last=
	Line 6117: 06-21 19:40:15.981   497   512 I chromium: [INFO:user_script_set.cc(201)] UserScripts: Match name=alert1_doument-start url=https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165962&last=
	Line 6128: 06-21 19:40:16.798   497   512 I chromium: [INFO:url_pattern_set.cc(232)] UserScripts: URLPatternSet::MatchesURL true https://m.gamer.com.tw/forum/C.php?bsn=60076&page=&snA=7165962&last=`

I searched both keywords 'm.gamer'. If you looked at snA=id which indicate the change of post there are 5 uniques one for v100 but only 2 for v101(as I ensure total of 1+4 redirect in both case).  I am not a technical person but I would be more leaned towards what changed from 100.0.4896.135 that leads to this problem. However, I can only reproduce this behavior on 2 of my devices as I do not have more devices to test.

**Reply by n-buna404** [reply]
https://thumbsnap.com/3AsegzzM
Here is video demo to reproduce this issue for more clarity. Hope there are others that can help test this out. Note the audio might be quite low.
I changed the targeted url to https://www.google.com/
The script
`// ==UserScript==`
`// @name alert1_doument-start`
`// @version 0.1.00`
`// @description alert1`
`// @author mac`
`// @run-at document-end`
`// @match https://www.google.com/*`
`// ==/UserScript==`
`'use strict';`
`(() => alert('alert 1\nalert_doument-start'))();`
edit: I am sorry for accidentally tagging any unrelated users

**Reply by grant** [reply]
Don't @ me please

**Reply by csagan5** [reply]
> regarding the worker problem and the log bug, @csagan5 I will ask you to modify my patch with https://gist.github.com/uazo/c114f68d5c0c781a85045f958a8373da maybe after 103

@uazo you can submit the patch separately

**Comment by n-buna404** [comment]
For more visibility, I edited the not really relevant comment here.
https://thumbsnap.com/3AsegzzM
Here is video demo to reproduce this issue for more clarity. 
**I hope if this can reach out to Bromite users that willing to help reproduce the issue in the video. Note the audio might be quite low.**
Targeted url is https://www.google.com/ .

The script:

```js
// ==UserScript==
// @name test
// @match <all_urls>
// @run-at document-start
// ==/UserScript==
'use strict';
(() => alert('alert 1\nalert_document-start'))();
```

edit: I am sorry for accidentally tagging any unrelated users

**Comment by ncaat** [comment, marked as answer]
I'm having the same issue on version `104.0.5112.91`.
Steps to reproduce:
1. enable `chrome://flags/#enable-userscripts-log`
2. install this userscript:
    ```js
    // ==UserScript==
    // @name test
    // @match <all_urls>
    // @run-at document-start
    // ==/UserScript==
    document.documentElement.appendChild(document.createElement('style')).innerHTML='a{color:#f00 !important}';
    ```
3. go to https://en.m.wikipedia.org/wiki/Main_Page
4. go to https://en.m.wikipedia.org/wiki/Wikipedia
5. go to https://www.wikipedia.org/
6. repeat step 4
7. long press on the link of step 3 and select "Preview page"
8. repeat steps 4-6 in the preview

The userscript injection works normally and is logged for all steps except 4 where nothing related to userscripts is logged at all. It seems that the userscript is only injected on the first page of each origin and that visiting a different origin resets this behavior. Previewing a page doesn't have this issue. (Side note: `alert()` doesn't work in previews so you can't use that to test it)
@uazo I would really appreciate it if you could take another look at this issue.

**Reply by uazo** [reply]
> I would really appreciate it if you could take another look at this issue.

Certainly! but it is not reproducible... it works for me

**Reply by ncaat** [reply]
Thanks for the quick reply, weird that you can't reproduce it. Do you know what the difference in preview mode could be? Anything else I could do to debug this issue?

**Reply by n-buna404** [reply]
well using the steps above I can reproduce the issue on latest version (arm or arm64) but not 100.0.4896.135.
My device is Xiaomi Mi A1 with SD625 android 9,  Since i reinstall bromite fresh, the only change is adding the userscript. I did not touch any default setting. Though I am the one that originally raise the issue. We need to somehow find the cause of discrepancy. My deduction is that something changed from 100 to 101 that affects the userscript injection for same origin for some devices but it was strange that preview mode was not affected. Good find from ncaat

**Reply by uazo** [reply]
maybe I wrong the steps, maybe someone can provide a video?

**Reply by ncaat** [reply]
Video: https://anonfiles.com/z1o1H47ay9/screen-20220912-093415_mp4

**Reply by uazo** [reply]
thanks for the video, I did the steps correctly but it doesn't happen to me.

> The userscript injection works normally and is logged for all steps except 4 where nothing related to userscripts is logged at all. 

I think about some sort of cache or preload (which should be disabled in bromite) or maybe the page is built with javascript (ajax or similar). can you check with the devtools?

**Reply by n-buna404** [reply]
I would like to know about any settings that u changed that may cause the discrepancy. Even very small would be sufficient.
Another discrepancy observation is on github repo page example https://github.com/bromite/bromite. I can click issue, pull requests, discussion and even clicking into discussion but the red accent always stay but not for wikipedia or google and etc. however, using preview page all works perfectly.

**Reply by uazo** [reply]
> I would like to know about any settings that u changed that may cause the discrepancy

for me nothing, freshly installed clean bromite

I add here the test to try it with browserstack: [test5.js.txt](https://github.com/bromite/bromite/files/9546224/test5.js.txt)

**Reply by uazo** [reply]
with browserstack is reproducible...

**Reply by ncaat** [reply]
The issue is also present while javascript is disabled globally.

**Reply by uazo** [reply]
is the back-forward cache `chrome://flags/#back-forward-cache`, try to disable it
I'll verify

**Reply by ncaat** [reply]
Yep that solves it

**Reply by n-buna404** [reply]
can confirm disabling chrome://flags/#back-forward-cache makes my script works like expected when undergo same domain redirection

**Reply by ncaat** [reply]
@uazo could the injection mechanism maybe be changed to also inject into cached pages?

**Reply by uazo** [reply]
yes, I already found the bug

**Reply by n-buna404** [reply]
seems like the bug is back with version 106.05249.72
I tried disable back-forward cache `chrome://flags/#back-forward-cache` and the script works as expected again.
Anyone else can confirm?

**Reply by uazo** [reply]
yes, my fault, @csagan5 I didn't warn you (nor did I remember) that some of the changes I had separated for when I was going to make the pull.

https://github.com/uazo/bromite/blob/next-v106/build/patches/00Experimental-user-scripts-support--fixup-.patch
https://github.com/uazo/bromite/blob/next-v106/build/patches/00Experimental-user-scripts-support--fixup--2.patch

I will fix with 107.

**Reply by Arien02** [reply]
Just to remember, problem still persist with recent 106.0.5249.163.
